File: src/examples/ruby/webapp/order_mgr.rb
require('vertx')
require('json')

@eb = Vertx::EventBus

@eb.register_handler('demo.orderMgr') do |message|
  validateUser(message)
end

def validateUser(message)
  @eb.send('demo.authMgr.validate', { 'sessionID' => message.body['sessionID'] }) do |reply|
    if reply.body['status'] == 'ok'
      message.body['username'] = reply.body['username']
      saveOrder(message)
    else
      puts "Failed to validate user"
    end
  end
end

def saveOrder(message)
  @eb.send('demo.persistor',
      {'action' => 'save', 'collection' => 'orders', 'document' => message.body}) do |reply|
    if reply.body['status'] === 'ok'
      puts "order successfully processed!"
      message.reply({'status' => 'ok'}) # Reply to the front end
    else
      puts "Failed to save order"
    end
  end
end

File: src/examples/javascript/webapp/order_mgr.js
load('vertx.js');

var eb = vertx.eventBus;
var log = vertx.logger;

eb.registerHandler('demo.orderMgr', function(order, replier) {
  validateUser(order, replier);
});

function validateUser(order, replier) {
  eb.send('demo.authMgr.validate', { sessionID: order.sessionID }, function(reply) {
    if (reply.status === 'ok') {
      order.username = reply.username;
      saveOrder(order, replier);
    } else {
      log.warn("Failed to validate user");
    }
  });
}

function saveOrder(order, replier) {
  eb.send('demo.persistor',
      {action:'save', collection: 'orders', document: order}, function(reply) {
    if (reply.status === 'ok') {
      log.info("order successfully processed!");
      replier({status: 'ok'}); // Reply to the front end
    } else {
      log.warn("Failed to save order");
    }
  });
}

